Use tinyobjloader for loading .obj files

stonier opened the pull request:
This follows up on @nkoenig's pull request #4832.

* Collapses into one commit
* Fixes an indexing bug in #4382 which dropped the last vertex
* Resolves trivial conflict errors
* Makes sure the [scale fix](https://github.com/RobotLocomotion/drake/commit/203a9a83dcd69a2a964230a4e6e071ac3f9e6224#diff-581f55e152daf186e6be7b725b9257c2R330) (added after #4832) is captured in the new PR
* Makes sure the [complex shape checking](https://github.com/RobotLocomotion/drake/commit/66a68229e8ae687bc7f8194524dafbed0ed2f4e0#diff-581f55e152daf186e6be7b725b9257c2R323) (added after #4832) is captured in the new PR

For conflicts, had to mostly make sure the commits to [drake/multibody/shapes/geometry.cc](https://github.com/RobotLocomotion/drake/commits/master/drake/multibody/shapes/geometry.cc) after Jan 24 were incorporated.

Some comments:

TinyObjLoader can do 'on the fly' triangulation. However it doesn't do any checking if the triangulation is correct as was added in 66a68229. It's additionally impossible as far as I can tell to check for such since you do not have the list of non-triangular faces that were triangulated after the fact. Subsequently, triangulation is off by default and most of the original checking code for complex shapes has been left in. With tinyobjloader, you still get the benefits of better parsing (for instance the parser order error problem is no longer a problem - see the test changes below). It might be worth adding triangulation checks upstream.

Testing:

Testing was mostly resolved via the bazel `mesh_triangulate_test` in `drake/multibody/shapes`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5631)
<!-- Reviewable:end -->

stonier commented:
Decided to forge on ahead without Nate on this one. Useful learning experience for myself and better for Nate to tackle current problems than have to rewind his brain and handle conflict resolutions due to the lengthy delay.

stonier commented:
I think +@liangfok for feature review since originally tasked with it on @nkoenig's original PR (#4832).

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - I left a bunch of minor nits. Thanks for doing this!

@seancurtis-tri, can you double-check the modifications to `drake/multibody/shapes/`?

---

Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[WORKSPACE, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHhdbBMCWHfhNIuH8u:-KgHhdbBMCWHfhNIuH8v:b7908j7) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/WORKSPACE#L158)):*
> ```Python
>     name = "tinyobjloader",
>     remote = "https://github.com/syoyo/tinyobjloader.git",
>     commit = "7c3206f9193ca259b73fec6556cdcab45c1f5723",
> ```

BTW, can we use a newer SHA? This one is from Dec 31, 2016.

---

*[drake/multibody/shapes/geometry.cc, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHdmdYrWtblfE8zz3G:-KgHdmdYrWtblfE8zz3H:bm8hbzp) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L17)):*
> ```cpp
> using std::string;
> using std::ifstream;
> using std::istringstream;
> ```

BTW: nit: alpha sort.

---

*[drake/multibody/shapes/geometry.cc, line 313 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHe2LkKbA-1cJe1gTV:-KgHe2LlkA1IAvYjXsBW:b-42xbdl) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L313)):*
> ```cpp
> 
>   string path;
>   size_t idx = obj_file_name.rfind('/');
> ```

BTW, can `size_t` be `const size_t`?

---

*[drake/multibody/shapes/geometry.cc, line 323 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHhu269_6noX0Pz26U:-KgHhu269_6noX0Pz26V:bygy1xz) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L323)):*
> ```cpp
>   std::string err;
> 
>   // Would be nice to hand off triangulation for non-triangular faces
> ```

BTW, should "for" be "of"?

---

*[drake/multibody/shapes/geometry.cc, line 338 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHeZ3nvcufqSvdRHmj:-KgHeZ3nvcufqSvdRHmk:bb16mrb) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L338)):*
> ```cpp
> 
>   // Store the vertices.
>   for (size_t index = 0; index < attrib.vertices.size() ; index += 3) {
> ```

For this application, we typically require the use of `int` rather than `size_t`. It will require using `static_cast<int>(attrib.vertices.size())`.

https://google.github.io/styleguide/cppguide.html#Integer_Types

---

*[drake/multibody/shapes/geometry.cc, line 344 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHf7ibDQ9KSY-x6EW7:-KgHf7ic1J2opJ9jFofu:b-z9905c) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L344)):*
> ```cpp
>   }
> 
>   // Counter for triangles added to compensate for non-triangular faces
> ```

BTW, a period is missing at end of this sentence.

https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

---

*[drake/multibody/shapes/geometry.cc, line 346 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHfJqPbcUvqx19lNTT:-KgHfJqPbcUvqx19lNTU:b370565) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L346)):*
> ```cpp
>   // Counter for triangles added to compensate for non-triangular faces
>   int added_triangles = 0;
>   // Counter keeping track of indices, used to check against the #vertices above
> ```

BTW, a period is missing at end of this sentence.

https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

Also, we typically don't allow abbreviations like "#vertices".

https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

---

*[drake/multibody/shapes/geometry.cc, line 351 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgN1eYkZzMCuvn5ND:-KgHgN1eYkZzMCuvn5NE:b-ixg5vx) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L351)):*
> ```cpp
>   // Iterate over the shapes.
>   for (auto const& shape : shapes) {
>     unsigned int index_offset = 0;
> ```

`unsigned int` should be `int` per style guide.

https://google.github.io/styleguide/cppguide.html#Integer_Types

---

*[drake/multibody/shapes/geometry.cc, line 354 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHfpyA9zXvvezBdMlN:-KgHfpyA9zXvvezBdMlO:b-ixg5vx) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L354)):*
> ```cpp
> 
>     // For each face in the shape.
>     for (unsigned int face = 0;
> ```

`unsigned int` should be `int` per style guide.

https://google.github.io/styleguide/cppguide.html#Integer_Types

---

*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgHgBrq8MWKnrIVWcTP:b7d030h) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
> ```cpp
>     for (unsigned int face = 0;
>          face < shape.mesh.num_face_vertices.size(); ++face) {
>       const size_t vert_count = shape.mesh.num_face_vertices[face];
> ```

`size_t` should be `int` per style guide.

https://google.github.io/styleguide/cppguide.html#Integer_Types

---

*[drake/multibody/shapes/geometry.cc, line 359 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHg3uEc5Ei7nXugDw9:-KgHg3uEc5Ei7nXugDwA:b7d030h) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L359)):*
> ```cpp
> 
>       std::vector<int> indices;
>       for (size_t vert = 0; vert < vert_count; ++vert) {
> ```

`size_t` should be `int` per style guide.

https://google.github.io/styleguide/cppguide.html#Integer_Types

---

*[drake/multibody/shapes/geometry.cc, line 367 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgW8JmGCKKWp1vcdn:-KgHgW8K5llCtskwkfM1:b-z9905c) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L367)):*
> ```cpp
> 
>       // Make sure the face has three vertices. This can only occur if
>       // do_tiny_obj_triangulation is false and faces are non triangular
> ```

BTW, a period is missing at end of this sentence.

https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

---

*[drake/multibody/shapes/test/mesh_triangulate_test.cc, line 152 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHh7TWqFO9dR66GbVT:-KgHh7TWqFO9dR66GbVU:b-z9905c) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/test/mesh_triangulate_test.cc#L152)):*
> ```cpp
> // Tests the triangulation code's response when a face references a vertex
> // not yet written in the file. A simple parser may be thrown by this, but
> // a good parser should handle it (e.g. tinyobjloader does)
> ```

BTW, a period is missing at end of this sentence.

https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgHhHEyPSZk8TwTIcyo:b-4833am)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[WORKSPACE, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHhdbBMCWHfhNIuH8u:-KgJufXMqxGA8XD7PdR0:b-11z5bo) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/WORKSPACE#L158)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, can we use a newer SHA? This one is from Dec 31, 2016.
</blockquote></details>

Looks like Nate was using a different approach in the WORKSPACE file as well (new_git_repository vs github_archive). I'll look into that as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKCA0awvJOw2Nw4Wwd:btwdcwq) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

`size_t` should be `int` per style guide.

https://google.github.io/styleguide/cppguide.html#Integer_Types
</blockquote></details>

They actually recommend the use of size_t where appropriate. Is that not taken on board? In this case though, it really would be std::vector<int>::size_type (or auto) that would be the appropriate choice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 344 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHf7ibDQ9KSY-x6EW7:-KgKChoz9Off5UHUEVE_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L344)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, a period is missing at end of this sentence.

https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 9 of 11 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKDUVlMoMg2zAgnYZn:bwuja0p) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
Good question. I am not 100% sure in this case. Here is Drake's refinement to the rule:

http://drake.mit.edu/code_style_guide.html#clarifications

> When using Integer Types within Drake, unsigned types are forbidden, with the following exceptions (per #2514):
>
>    - uint32_t or uint64_t are allowed for bitfields, and
>    - size_t is allowed when a (1) non-Drake API uses unsigned types, and (2) casting to a signed type during Drake’s interactions with the non-Drake API would obscure the readability of our code, and (3) subtraction underflow below zero is obviously not at risk.

I'll mark myself satisfied and allow the platform reviewer to revisit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKDuhD_y6spjXZaYuA:b-lyy630) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Good question. I am not 100% sure in this case. Here is Drake's refinement to the rule:

http://drake.mit.edu/code_style_guide.html#clarifications

> When using Integer Types within Drake, unsigned types are forbidden, with the following exceptions (per #2514):
>
>    - uint32_t or uint64_t are allowed for bitfields, and
>    - size_t is allowed when a (1) non-Drake API uses unsigned types, and (2) casting to a signed type during Drake’s interactions with the non-Drake API would obscure the readability of our code, and (3) subtraction underflow below zero is obviously not at risk.

I'll mark myself satisfied and allow the platform reviewer to revisit.

</blockquote></details>

I suspect that this code fails point (3) -- _"subtraction underflow below zero is obviously not at risk."_ -- because it's not trivially obvious that there is no subtraction, because the lifetime of these variables exceeds one pageful of code.  If the methods were shorter, we might allow it, but I think in this case we need `int`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKI8SEhEB11jIFADrz:b-zbhbn0) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I suspect that this code fails point (3) -- _"subtraction underflow below zero is obviously not at risk."_ -- because it's not trivially obvious that there is no subtraction, because the lifetime of these variables exceeds one pageful of code.  If the methods were shorter, we might allow it, but I think in this case we need `int`.
</blockquote></details>

Would you not be reducing that risk, but introducing the risk that you're casting to a differently sized type with potential overflow problems?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 367 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgW8JmGCKKWp1vcdn:-KgKKx3H-3Pk0u1wu_ml:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L367)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, a period is missing at end of this sentence.

https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 346 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHfJqPbcUvqx19lNTT:-KgKL-K7vtH23glOT1oh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L346)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, a period is missing at end of this sentence.

https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

Also, we typically don't allow abbreviations like "#vertices".

https://google.github.io/styleguide/cppguide.html#General_Naming_Rules
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 323 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHhu269_6noX0Pz26U:-KgKL2JZK76QJFx2tMxk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L323)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should "for" be "of"?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 313 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHe2LkKbA-1cJe1gTV:-KgKL3VFB6qucWT0FJ2m:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L313)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, can `size_t` be `const size_t`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHdmdYrWtblfE8zz3G:-KgKL9U-O3sn3rEkJVUo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L17)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: nit: alpha sort.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[WORKSPACE, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHhdbBMCWHfhNIuH8u:-KgKNF4KA1aCYYKQ7sXB:bkfxr5a) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/WORKSPACE#L158)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Looks like Nate was using a different approach in the WORKSPACE file as well (new_git_repository vs github_archive). I'll look into that as well.
</blockquote></details>

From what I can gather - the bazel build does not require the git submodules - that was purely for the cmake. Is this correct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[WORKSPACE, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHhdbBMCWHfhNIuH8u:-KgKOCvVEXyJSGtYQKVm:b-ocvm8k) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/WORKSPACE#L158)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

From what I can gather - the bazel build does not require the git submodules - that was purely for the cmake. Is this correct?
</blockquote></details>

Yes. This is correct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKOUgjOFwabPCNAVjB:b-s4h9d) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Would you not be reducing that risk, but introducing the risk that you're casting to a differently sized type with potential overflow problems?
</blockquote></details>

I agree there is an increased overflow risk using a signed integer, but since the values we're dealing with are closer to zero, I think the risk of an unsigned number underflowing is larger than a signed number overflowing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKSCJnBOMAhMIFNQW8:b339lys) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I agree there is an increased overflow risk using a signed integer, but since the values we're dealing with are closer to zero, I think the risk of an unsigned number underflowing is larger than a signed number overflowing.
</blockquote></details>

I guess your history defines you. I've been burned very often by code casting to be deployed embedded boards where type sizes change with each board. :)

Switched them all to ints here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 9 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKSo-oGcwQRL1cekeW:b-cqpcot) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

I guess your history defines you. I've been burned very often by code casting to be deployed embedded boards where type sizes change with each board. :)

Switched them all to ints here.
</blockquote></details>

That's true. More care will definitely be needed should we decide to add embedded hardware details to Drake's [matrix of supported configurations](http://drake.mit.edu/developers.html#supported-configurations).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 3 files at r4.
Review status: 10 of 11 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 3 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKTVeMOE8xVygPauEK:b-k1zre7) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

That's true. More care will definitely be needed should we decide to add embedded hardware details to Drake's [matrix of supported configurations](http://drake.mit.edu/developers.html#supported-configurations).
</blockquote></details>

Drake is probably pretty safe in that regard...can't envisage this getting onto an armv6.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 356 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgBrq8MWKnrIVWcTO:-KgKTaRteFWZ1FToF7Nl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L356)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Drake is probably pretty safe in that regard...can't envisage this getting onto an armv6.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 359 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHg3uEc5Ei7nXugDw9:-KgKTc93BQPDL0IcLAyy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L359)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

`size_t` should be `int` per style guide.

https://google.github.io/styleguide/cppguide.html#Integer_Types
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 354 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHfpyA9zXvvezBdMlN:-KgKTeRZSQ7nfrw_OmTD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L354)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

`unsigned int` should be `int` per style guide.

https://google.github.io/styleguide/cppguide.html#Integer_Types
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 351 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHgN1eYkZzMCuvn5ND:-KgKTfatu1Hf8IVr6fEE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L351)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

`unsigned int` should be `int` per style guide.

https://google.github.io/styleguide/cppguide.html#Integer_Types
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[drake/multibody/shapes/geometry.cc, line 338 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHeZ3nvcufqSvdRHmj:-KgKU4g_vnHeAC8D3hBP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/drake/multibody/shapes/geometry.cc#L338)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

For this application, we typically require the use of `int` rather than `size_t`. It will require using `static_cast<int>(attrib.vertices.size())`.

https://google.github.io/styleguide/cppguide.html#Integer_Types
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[WORKSPACE, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHhdbBMCWHfhNIuH8u:-KgKUS37grJgjMKLs1Uo:bucgevc) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/WORKSPACE#L158)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Yes. This is correct.
</blockquote></details>

Switched to the latest commit, shifted to github_archive and updated the submodule. Tested.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
Nits picked. Just awaiting a check from @seancurtis-tri.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgKV08cFGX1ZLxooImm:b9l68yz)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[WORKSPACE, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgHhdbBMCWHfhNIuH8u:-KgKVDmL4Av33R5yyi_s:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02f1fd1dc9a42f8d38478ac0d5f6387844cd34a4/WORKSPACE#L158)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Switched to the latest commit, shifted to github_archive and updated the submodule. Tested.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Daniel, great to see you on here! A Reviewable tip: each time you hit "Publish" in Reviewable email notifications get sent out to basically everyone. (I've gotten 34 emails so far on this PR, mostly containing just "Done.") Generally it is better to respond to many comments before hitting Publish so that one notification is sent. Reviewable is *very* good about not losing anything you've done, almost down to individual characters typed.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgKq33sx81aUcABXNaW:b5ijf1q)*
<!-- Sent from Reviewable.io -->

stonier commented:
Aye, I went hunting Liang earlier to get a tutorial on Reviewable and discovered the rightmost button (immediate send) was not necessary. Interesting how compulsive finalising a task with a button click is ... I just don't get the same feeling of satisfaction with a no-click finish! Apologies for the spam.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLaK-vvfpE_AmX_vxu:-KgLaK-w0hXC7uqCLV-V:bq11ftp) (no related file):*
+@sammy-tri for platform review

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgKzfYpl2EKPnV4W76h:b-ki73si)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 7 of 11 files at r1, 1 of 2 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLaK-vvfpE_AmX_vxu:-KgLfJ3-GQlx7yrL63cm:b-s794vx) (no related file):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

+@sammy-tri for platform review
</blockquote></details>

+(status: curate commits before merging) 

---

*[drake/multibody/shapes/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLfs2GKs9bTQs81LWb:-KgLfs2HffBVV7yV8DXj:bzcskdi) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/CMakeLists.txt#L6)):*
> ```CMake
>     Eigen3::Eigen
>     spruce
>     tinyobjloader)
> ```

If tinyobjloader is going to be optional in cmake, do we need an if(tinyobjloader_FOUND) around targets which use it?  (I actually don't know the answer for sure, and I'm not sure how much attention the cmake build needs at this point)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+@jwnimmer-tri <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for extra build system review.

Once this PR review completes, please opt-in to an OSX build before merging -- because the PR contains changes to the third-party external libraries, it's at relatively high risk for breaking on OSX versus a more typical PR.  Just post a github comment with `@drake-jekins-bot mac-clang-ninja-experimental please` (not in backticks).

---

Reviewed 7 of 11 files at r1, 2 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/shapes/geometry.cc, line 7 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLsoeuNNSIuXMdZRPA:-KgLsoevQQQ2ZijGSQc6:bazyxs2) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/geometry.cc#L7)):*
> ```cpp
> #include <fstream>
> #include <stdexcept>
> #include <tiny_obj_loader.h>
> ```

BTW This include statement should be grouped with spruce.hh (not stdexcept) per https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes because it's a third-party library (group 4), not system (group 3).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgLt7GaMtY-uoY2j8fz:b-ejito1)*
<!-- Sent from Reviewable.io -->

stonier commented:
@drake-jekins-bot mac-clang-ninja-experimental please

stonier commented:
@drake-jenkins-bot mac-clang-ninja-experimental please

liangfok commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
Bingo - boom.  From [jenkins/mac-clang-ninja-experimental#244](https://drake-jenkins.csail.mit.edu/job/mac-clang-ninja-experimental/244/consoleFull):

```
dyld: Library not loaded: libtinyobjloader.1.dylib
```

Who is an expert on the macos side of the build? 

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLfs2GKs9bTQs81LWb:-KgP7k5jMjPYFrkatsmY:b-nydigx) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/CMakeLists.txt#L6)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

If tinyobjloader is going to be optional in cmake, do we need an if(tinyobjloader_FOUND) around targets which use it?  (I actually don't know the answer for sure, and I'm not sure how much attention the cmake build needs at this point)
</blockquote></details>

It's not using the cmake find_package mechanism, so tinyobjloader_FOUND is not a valid call. It is however going through the drake_add_external mechanism which supplies a WITH_TINYOBJLOADER option. 

This is where it gets a little unintuitive. Right now even non-optional externals have this option available (e.g. bullet, yaml-cpp). There are some notes in cmake/externals.cmake about having these around so they can be skipped in CI cpplint jobs so if someone disables these when drake is being compiled, then it looks as if the compile will break. 

Long story short. Tinyobjloader must be on for the compile. Nate I believe was following the existing infrastructure already there which is at least more consistent than doing something different. The alternative is to unwind the cmake logic for everything so this isn't confusing....or unwind cmake builds entirely, which is looming.

---

*[drake/multibody/shapes/geometry.cc, line 7 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLsoeuNNSIuXMdZRPA:-KgPFSnZs6PUUAZWfYb5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/geometry.cc#L7)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This include statement should be grouped with spruce.hh (not stdexcept) per https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes because it's a third-party library (group 4), not system (group 3).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgPrpycljVFp36l6y08:b-5kmw2x)*
<!-- Sent from Reviewable.io -->

patmarion commented:
Just curious, have you tried the reader for each .obj file in the drake repo?  Maybe the urdf parser tests covers them all, but I'm not sure.

jwnimmer-tri commented:
> Who is an expert on the macos side of the build?

David usually bites the bullet and fixes OS X for us; sometimes Soonho helps as well (and he can help you get a login to an in-office OS X test machine).

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgPti2X6rXii0qgdbvM:bs0qmec)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
@stonier , please DM me in Slack if you need an account in our mac server.

I couldn't reproduce the problem on my mac using bazel. I just triggered cmake build and am waiting for its outcome.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgPwefRt2ppHL_ewWi3:b-bn6utq)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Using cmake, I can reproduce the problem on my mac.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgQ5_sXGvbe6pPXhYXI:b-fyaqdc)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
`otool -L drake/./multibody/collision/test/collision_filter_group_test` gives me the following:
```
drake/./multibody/collision/test/collision_filter_group_test:
...
	@rpath/libdrakeShapes.dylib (compatibility version 0.0.0, current version 0.0.0)
	libtinyobjloader.1.dylib (compatibility version 1.0.0, current version 1.0.4)
...
```
I think it should have `@rpath/libtinyobjloader.1.dylib` instead of `libtinyobjloader.1.dylib`.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgQ5zjDqD_iGkqr-iK7:b-ggzuhj)*
<!-- Sent from Reviewable.io -->

patmarion commented:
I believe this would be resolved if this PR were merged:

https://github.com/RobotLocomotion/drake/pull/4830

That PR introduces CMake code that forces external projects (tinyobjloader, in this case) to install libraries with RPATH in their install name.  Implemented by this commit:

https://github.com/RobotLocomotion/drake/pull/4830/commits/f3b440547ad40cef0355123694562ab1ead217e0

It is also possible to fix the issue directly in tinyobjloader's cmake build by enabling the `CMAKE_MACOSX_RPATH` option.

sammy-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLfs2GKs9bTQs81LWb:-KgQ9OqWSU2ET6mYogp2:b1zsot2) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/CMakeLists.txt#L6)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

It's not using the cmake find_package mechanism, so tinyobjloader_FOUND is not a valid call. It is however going through the drake_add_external mechanism which supplies a WITH_TINYOBJLOADER option. 

This is where it gets a little unintuitive. Right now even non-optional externals have this option available (e.g. bullet, yaml-cpp). There are some notes in cmake/externals.cmake about having these around so they can be skipped in CI cpplint jobs so if someone disables these when drake is being compiled, then it looks as if the compile will break. 

Long story short. Tinyobjloader must be on for the compile. Nate I believe was following the existing infrastructure already there which is at least more consistent than doing something different. The alternative is to unwind the cmake logic for everything so this isn't confusing....or unwind cmake builds entirely, which is looming.
</blockquote></details>

I actually used yaml-cpp as the example and searched for its users in the drake build.  It does seem to be protected by yaml-cpp_FOUND  (and, sure, the FOUND spelling may not be appropriate for this case).  There's a lot of Bullet_FOUND floating around the CMakeLists.txt as well.   Are you saying those are superfluous and could/should be removed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/multibody/shapes/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLfs2GKs9bTQs81LWb:-KgQCMBH-hgb3uJ9Kbts:b54jqrt) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/CMakeLists.txt#L6)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I actually used yaml-cpp as the example and searched for its users in the drake build.  It does seem to be protected by yaml-cpp_FOUND  (and, sure, the FOUND spelling may not be appropriate for this case).  There's a lot of Bullet_FOUND floating around the CMakeLists.txt as well.   Are you saying those are superfluous and could/should be removed?
</blockquote></details>

I believe checking `tinyobjloader_FOUND` (or some variation of that) is correct because `WITH_TINYOBJLOADER` is simply asking CMake to "please checkout then find the package". Whether the package was actually found can be determined by checking the corresponding `_FOUND` variable. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/multibody/shapes/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLfs2GKs9bTQs81LWb:-KgQEZIvpF5d2iwGtOY3:b7ej6rk) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/CMakeLists.txt#L6)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I believe checking `tinyobjloader_FOUND` (or some variation of that) is correct because `WITH_TINYOBJLOADER` is simply asking CMake to "please checkout then find the package". Whether the package was actually found can be determined by checking the corresponding `_FOUND` variable. 
</blockquote></details>

I instrumented this CMakeLists.txt to print the variable names and here are the ones related to `tinyobjloader`:

```
-- tinyobjloader_CONFIG=/home/liang/dev/drake-distro-1/build/install/lib/tinyobjloader/cmake/tinyobjloader-config.cmake
-- tinyobjloader_CONSIDERED_CONFIGS=/home/liang/dev/drake-distro-1/build/install/lib/tinyobjloader/cmake/tinyobjloader-config.cmake
-- tinyobjloader_CONSIDERED_VERSIONS=1.0.4
-- tinyobjloader_DIR=/home/liang/dev/drake-distro-1/build/install/lib/tinyobjloader/cmake
-- tinyobjloader_FOUND=1
-- tinyobjloader_VERSION=1.0.4
-- tinyobjloader_VERSION_COUNT=3
-- tinyobjloader_VERSION_MAJOR=1
-- tinyobjloader_VERSION_MINOR=0
-- tinyobjloader_VERSION_PATCH=4
-- tinyobjloader_VERSION_TWEAK=0
-- tinyxml2_LIB_DEPENDS=
```

Looks like, this library should be conditioned on `tinyobjloader_FOUND`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/multibody/shapes/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLfs2GKs9bTQs81LWb:-KgQQmrR3R00zI_GRY0c:b-q9atmu) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/CMakeLists.txt#L6)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I instrumented this CMakeLists.txt to print the variable names and here are the ones related to `tinyobjloader`:

```
-- tinyobjloader_CONFIG=/home/liang/dev/drake-distro-1/build/install/lib/tinyobjloader/cmake/tinyobjloader-config.cmake
-- tinyobjloader_CONSIDERED_CONFIGS=/home/liang/dev/drake-distro-1/build/install/lib/tinyobjloader/cmake/tinyobjloader-config.cmake
-- tinyobjloader_CONSIDERED_VERSIONS=1.0.4
-- tinyobjloader_DIR=/home/liang/dev/drake-distro-1/build/install/lib/tinyobjloader/cmake
-- tinyobjloader_FOUND=1
-- tinyobjloader_VERSION=1.0.4
-- tinyobjloader_VERSION_COUNT=3
-- tinyobjloader_VERSION_MAJOR=1
-- tinyobjloader_VERSION_MINOR=0
-- tinyobjloader_VERSION_PATCH=4
-- tinyobjloader_VERSION_TWEAK=0
-- tinyxml2_LIB_DEPENDS=
```

Looks like, this library should be conditioned on `tinyobjloader_FOUND`.
</blockquote></details>

The guard condition should be added only if `tinyobjloader` is an optional external.  If you look at `packages.cmake` above, you will see that `tinyobjloader` is marked as `REQUIRED`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
@drake-jenkins-bot mac-clang-ninja-experimental please

liangfok commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
@patmarion thanks, spot on.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLaK-vvfpE_AmX_vxu:-KgReCzbqbg7XimRAKkL:bb6un1j) (no related file):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

+(status: curate commits before merging) 
</blockquote></details>

Done

---

*[drake/multibody/shapes/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgLfs2GKs9bTQs81LWb:-KgQk-I41sU7dYVS6syc:b8g7fcm) ([raw file](https://github.com/robotlocomotion/drake/blob/736a0db659b68280bbdcbf77c84fa68a659c2e7b/drake/multibody/shapes/CMakeLists.txt#L6)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The guard condition should be added only if `tinyobjloader` is an optional external.  If you look at `packages.cmake` above, you will see that `tinyobjloader` is marked as `REQUIRED`.
</blockquote></details>

Ah, I see why I missed the FOUND calls before - was grepping from the build directory. My bad. Thanks for the pointer jeremy.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgRe7keW-CWfKAPxLx6:ba5i3do)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
@drake-jenkins-bot mac-clang-ninja-experimental please

liangfok commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

stonier commented:
Working with the latest PR from @patmarion. @sammy-tri should be all good now.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgUW2vz8yzFC1gOx3vF:b73cgnk)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
There's still an unresolved question around the _FOUND tests, namely: are the tests for yaml-cpp_FOUND which are already in the tree incorrect/unnecessary, or does this change require similar guards?  Or are we going to punt on the question because CMake is deprecated infrastructure?

---

Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgVIvgyWCMwBDoJ2Iqr:b-39berx)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
tinyobjloader is required, so needs no guards.  yaml-cpp is optional, so needs guards.

sammy-tri commented:
This is not obvious to me, because the syntax declaring them in options.cmake appears to be identical.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgVKvJ4-sVJ93E9WrCx:byycm04)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
It's in `packages.cmake`.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgVLH2ZHqS79y8ER2tq:b-l9bpcp)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Oh, good point.  `tinyobjloader` is required in one place (packages.cmake) and optional in another (options.cmake).  They should both be required.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgVMNuEr7MCMrlj5eNQ:b-ul0new)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgVMEvufz-lQ9dGyn8m:bnfp4nl)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[cmake/options.cmake, line 260 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgVMtvqndFV89ottZiQ:-KgVMtvsaHHQz_1_JZO3:bg1idoc) ([raw file](https://github.com/robotlocomotion/drake/blob/c83039487225fe0179f8c3411acbb41b37b9c583/cmake/options.cmake#L260)):*
> ```CMake
> 
>   drake_system_dependency(
>     TINYOBJLOADER OPTIONAL REQUIRES tinyobjloader
> ```

Use of the OPTIONAL keyword here seems to contradict the package being required in packages.cmake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
(pending the one open discussion, that is)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-:-KgVN8QbZgqrdZsq7MKR:b-sulbh9)*
<!-- Sent from Reviewable.io -->

stonier commented:
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[cmake/options.cmake, line 260 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/5631#-KgVMtvqndFV89ottZiQ:-KgVoSqQan31ygitOOCi:b-sx0b7w) ([raw file](https://github.com/robotlocomotion/drake/blob/c83039487225fe0179f8c3411acbb41b37b9c583/cmake/options.cmake#L260)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Use of the OPTIONAL keyword here seems to contradict the package being required in packages.cmake.
</blockquote></details>

Thanks, good catch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5631)*
<!-- Sent from Reviewable.io -->
